src/osd/OSD.cc

```diff
@@ -6089,6 +6071,11 @@ void OSD::do_waiters()
   assert(osd_lock.is_locked());
 
   dout(10) << "do_waiters -- start" << dendl;
+  // take_waiters be called in the status of osd_lock is locked.
+  // And !finished.empty() is very rare situation.
+  if (finished.empty())
```

Comment:
Nothing I have found suggests that std::list::empty() is safe to call with a racing write.  I think this probably invites a rare crash.